### PR TITLE
High-level steps for citus warp

### LINKED ISSUE #326: Document migration via Citus Warp
## Why are we implementing it? (sales eng)

Citus Warp is an efficient, reliable way to populate a coordinator node prior to distributing data to workers. Alternatives like copying to/from CSV require more disk space and are less efficient.

### What are the typical use cases?

Migrating from RDS, or other Postgres installations to Citus.

### Are there relevant blog posts or outside documentation about the concept/feature?

There's a private google document with steps and screenshots.

We released a blog post about Citus Warp

### PR DESCRIPTION
**Pull request by begriffs**
Fixes #326 

In addition to the content in here, there's a question of its position in the menu. I put it next to our other migration items, but maybe it should be moved into their content rather than after them.

### CODE REVIEW

File: migration/data.rst [hunk 1]
```diff
@@ -0,0 +1,71 @@
+.. _data_migration:
+
+Data Migration
+==============
+
+Migrating from one database to another has traditionally been difficult. There are two traditional approaches: using a dump and restore that takes downtime, or updating application logic to write to two databases and then switch to the new one.
+
+Now there's an easier way. PostgreSQL 9.4 added logical replication to stream data to a new database even while the existing database continues to sustain application load. It's as if the application automatically writes to two databases rather than one, except with perfect transactional logic.
```
Comment (lfittl) [inline migration/data.rst]: `PostgreSQL 9.4 added logical replication` is not correct - Postgres 9.4 added logical *decoding* which can be used for replication, but the database itself does not provided that functionality. Postgres 10 has added actual logical replication into Postgres.

File: migration/data.rst [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+On RDS, edit the inbound database security group to add a custom TCP rule:
+
+Protocol
+  TCP
+Port Range
+  5432
+Source
+  <citus ip>/32
+
+This white-lists the IP address of the Citus coordinator node to make an inbound connection. An alternate way to connect the two is to establish peering between their virtual private clouds. We can help set that up if desired.
```
Comment (lfittl) [inline migration/data.rst]: I would call this `between their VPCs` (usually VPC is not spelled out)

File: migration/data.rst [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Put a freeze on the structure of the source database, telling application developers not to modify it. Mismatched table columns/types can confuse the replication process.
+
+5. Run the Warp
+~~~~~~~~~~~~~~~
+
+Contact us and a Cloud engineer will connect to your database with Citus Warp to create a basebackup, open a replication slot, and begin the replication. We can include/exclude your choice of tables in the migration.
+
+During the first stage, creating a basebackup, the Postgres write-ahead log (WAL) may grow substantially if the database is under write load. Make sure you have at least 20% free space on disk during this process. Once the backup is complete and replication begins then the database will be able to shrink the WAL again.
```
Comment (lfittl) [inline migration/data.rst]: maybe say `make sure you have sufficient disk space on the source database during this process. As a basic measurement 100GB or 20% of total disk space, whichever is greater, is a good metric in terms of how much space should be free before starting this process.`

File: migration/data.rst [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+When the replication has caught up with the current state of the source database, the application is free to connect to the new database instead. Any stray updates from the source database will continue to be reflected in the destination.
```
Comment (lfittl) [inline migration/data.rst]: "Whilst technically possible, we do not recommend writing to both the source and destination database at the same time."

File: migration/data.rst [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+Finally, when the application is pointing to the new database and no further changes are happening on the source database, contact us again to remove the replication slot. The migration is complete.
```
Comment (lfittl) [inline migration/data.rst]: I realize I missed a step in our chat today.

Due the the nature of the replication process, sequence values don't get updated correctly on the destination databases. In order to have the correct sequence value for e.g. an id column, you need to manually adjust the sequence values before turning on writes on the destination database. @saicitus should have a script for that, I believe.

**Comment by begriffs** [comment]
Good to go?

File: migration/data.rst [hunk 6]
```diff
@@ -0,0 +1,73 @@
+.. _data_migration:
+
+Data Migration
+==============
+
+Migrating from one database to another has traditionally been difficult. There are two traditional approaches: using a dump and restore that takes downtime, or updating application logic to write to two databases and then switch to the new one.
+
+Now there's an easier way. PostgreSQL 10 added logical replication to stream data to a new database even while the existing database continues to sustain application load. It's as if the application automatically writes to two databases rather than one, except with perfect transactional logic.
```
Comment (lfittl) [inline migration/data.rst]: People might be confused and think that we require Postgres 10 on the source - can we rewrite this to be less focused on the Postgres 10 logical replication?

For context, Citus Warp uses logical decoding, which is part of Postgres starting 9.4.

File: migration/data.rst [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+We have streamlined logical replication on :ref:`Citus Cloud <cloud_overview>` with a process called Citus Warp. To do a warp, we connect the coordinator node of a Citus cluster to an existing database through VPC peering or IP white-listing, and begin replication.
```
Comment (lfittl) [inline migration/data.rst]: "To do a warp," => "As part of the Citus Warp process,"

File: migration/data.rst [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+Using Citus Warp
+----------------
+
+Here's the high level overview of the process:
```
Comment (lfittl) [inline migration/data.rst]: "Here are the steps you need to perform before starting the Citus Warp process"

File: migration/data.rst [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+1. Duplicate structure of the schema on a destination Citus cluster
+2. Enable logical replication in the source database
+3. Allow a network connection from Citus coordinator node to source
+4. Make sure no changes will happen to the schema of the source (modifying data is fine, but not structure)
+5. Contact us to run the warp
```
Comment (lfittl) [inline migration/data.rst]: I'm not a fan of the term of "running the warp" - can we find a better way to say this? (especially "warp" as a lower case noun seems odd)

File: migration/data.rst [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+1. Duplicate schema
+~~~~~~~~~~~~~~~~~~~
+
+The first step in migrating data to Citus is making sure that the schemas match exactly, at least for the tables you choose to migrate. One way to do this is by running ``pg_dump --schema-only`` against the source database. Replay the output on the coordinator Citus node. Another way to is to run application migration scripts against the destination database.
+
+All tables that you wish to migrate must have primary keys. The corresponding destination tables must have primary keys as well, the only difference being that those keys are allowed to be composite to contain the distribution column as well, as described in :ref:`mt_schema_migration`.
+
+Also be sure to distribute tables across the cluster prior to starting a warp so that the data doesn't have to fit on the coordinator node alone.
```
Comment (lfittl) [inline migration/data.rst]: Might be worth referencing `create_distributed_table` here just so its clear whats meant with "distribute tables"

File: migration/data.rst [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+2. Enable logical replication
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Some hosted databases such as Amazon RDS require enabling replication by changing a server configuration parameter. On RDS you will need to create a new parameter group, set ``rds.logical_replication = 1`` in it, then change set the parameter group as the active one. Applying the change requires a database server reboot, which can be scheduled for the next maintenance window.
```
Comment (lfittl) [inline migration/data.rst]: We should mention whats required on non-RDS databases, see https://github.com/citusdata/citus_warp#required-postgres-configuration-on-the-source

File: migration/data.rst [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+3. Open access for network connection
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In the Cloud console, identify the hostname (it ends in ``db.citusdata.com``). Dig the hostname to find its IP address:
+
+.. code-block:: bash
+
+  dig +short <hostname> A
+
+On RDS, edit the inbound database security group to add a custom TCP rule:
```
Comment (lfittl) [inline migration/data.rst]: Maybe say, "If you're using RDS"

File: migration/data.rst [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+  5432
+Source
+  <citus ip>/32
+
+This white-lists the IP address of the Citus coordinator node to make an inbound connection. An alternate way to connect the two is to establish peering between their VPCs. We can help set that up if desired.
+
+4. Freeze schema changes
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Put a freeze on the structure of the source database, telling application developers not to modify it. Mismatched table columns/types can confuse the replication process.
```
Comment (lfittl) [inline migration/data.rst]: I'm not sure if its clear here that you only need to do this once you're actually planning to cut over to Citus Cloud for production.

Maybe we can make this a separate part after "Run the warp"

File: migration/data.rst [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+5. Run the Warp
```
Comment (lfittl) [inline migration/data.rst]: "Run the Warp" sounds odd to me - can we maybe call this "Starting Citus Warp replication"

File: migration/data.rst [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+~~~~~~~~~~~~~~~
+
+Contact us and a Cloud engineer will connect to your database with Citus Warp to create a basebackup, open a replication slot, and begin the replication. We can include/exclude your choice of tables in the migration.
```
Comment (lfittl) [inline migration/data.rst]: "Contact us" => "Contact us by opening a support ticket in the Citus Cloud console"

File: migration/data.rst [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+
+During the first stage, creating a basebackup, the Postgres write-ahead log (WAL) may grow substantially if the database is under write load. Make sure you have sufficient disk space on the source database before starting this process. We recommend 100GB free or 20% of total disk space, whichever is greater. Once the backup is complete and replication begins then the database will be able to archive unused WAL files again.
+
+When the replication has caught up with the current state of the source database, there is one more thing to do. Due to the nature of the replication process, sequence values don't get updated correctly on the destination databases. In order to have the correct sequence value for e.g. an id column, you need to manually adjust the sequence values before turning on writes on the destination database.
+
+Once this is all complete, the application is ready to connect to the new database. We do not recommend writing to both the source and destination database at the same time, although it is technically possible.
```
Comment (lfittl) [inline migration/data.rst]: No need to mention that its technically possible, just say we don't recommend it.

File: migration/data.rst [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+
+5. Run the Warp
+~~~~~~~~~~~~~~~
+
+Contact us and a Cloud engineer will connect to your database with Citus Warp to create a basebackup, open a replication slot, and begin the replication. We can include/exclude your choice of tables in the migration.
+
+During the first stage, creating a basebackup, the Postgres write-ahead log (WAL) may grow substantially if the database is under write load. Make sure you have sufficient disk space on the source database before starting this process. We recommend 100GB free or 20% of total disk space, whichever is greater. Once the backup is complete and replication begins then the database will be able to archive unused WAL files again.
+
+When the replication has caught up with the current state of the source database, there is one more thing to do. Due to the nature of the replication process, sequence values don't get updated correctly on the destination databases. In order to have the correct sequence value for e.g. an id column, you need to manually adjust the sequence values before turning on writes on the destination database.
```
Comment (lfittl) [inline migration/data.rst]: The part about the sequence values should be a separate section that talks about "Switching your production system to Citus Cloud" (which also talks about the schema change freeze)

**Comment by lfittl** [comment]
@begriffs Apologies for the late review (I blame vacation!), added some comments - I also asked @craigkerstiens to review the initial paragraph (see email).

**Comment by lfittl** [comment]
@begriffs Well, so the point is that you can try out Citus Warp without imposing a schema freeze on everyone - it will just stop working after you make incompatible DDL changes.

So I would recommend calling out this limitation, and then having the recommendation to freeze schema changes as part of a production cutover only (and otherwise just ask us to restart the Warp process for you when you have incompatible DDL changes).

Hope that helps.

**Comment by craigkerstiens** [comment]
:shipit:

**Pull request merged**